### PR TITLE
Add next-checker only when not added

### PR DESCRIPTION
**Pull request by conao3**
Add next-checker only when not added.
This PR solves #1661.

### CODE REVIEW

**Comment by fmdkdd** [comment]
Thanks for the patch.

Ensuring there are no duplicates in `next-checker` sounds good, but what if the checker is added with another error level.  I.e, we should prevent this situation:
```
(error . emacs-lisp-checkdoc)
(warning . emacs-lisp-checkdoc)
```
as it makes little sense.

**Comment by conao3** [comment]
> as it makes little sense.

What is your intention...?
This situation is a very rare case and it makes little sense?

I cannot understand that an example makes sense. Does it make sure that warnings and errors reported by checkers are unified as warnings/errors?

**Comment by fmdkdd** [comment]
Sorry for not being clearer: your patch checks equality of checker using `memq`, but `flycheck-add-next-checker` can take the `next` argument as the form `(LEVEL . CHECKER)`, which will always fail `memq`.  So adding `(t . emacs-lisp-checkdoc)` as a next-checker to `emacs-lisp` adds a duplicate, which is not wanted.  Conversely, checkers in the `next-checkers` property can also be of the form `(LEVEL . CHECKER)`, so you would need to account for this case as well.

**Comment by conao3** [comment]
Fix to handle your mentioned situation.
Please tell me if you came up with the better local variable name.

**Comment by cpitclaudel** [comment]
Sorry for letting this patch slip through.  I think the right way to handle this is to create a separate function `flycheck-remove-next-checker`, and to call that from `flycheck-add-next-checker`.  Something like this:

```diff
---
 flycheck.el | 35 ++++++++++++++++++++++++++++++-----
 1 file changed, 30 insertions(+), 5 deletions(-)

diff --git a/flycheck.el b/flycheck.el
index c038c3f5..0ded6dcb 100644
--- a/flycheck.el
+++ b/flycheck.el
@@ -2280,12 +2280,19 @@ flycheck--verify-princ-checker
         (princ "\n"))))
   (princ "\n"))
 
+(defun flycheck--get-next-checker-symbol (next)
+  "Get the checker symmbol of NEXT checker.
+
+NEXT should be either a cons (NEXT-CHECKER . LEVEL) or a
+symbol."
+  (if (consp next) (cdr next) next))
+
 (defun flycheck-get-next-checkers (checker)
   "Return the immediate next checkers of CHECKER.
 
 This is a list of checker symbols.  The error levels of the
 `:next-checker' property are ignored."
-  (mapcar (lambda (c) (if (consp c) (cdr c) c))
+  (mapcar #'flycheck--get-next-checker-symbol
           (flycheck-checker-get checker 'next-checkers)))
 
 (defun flycheck-all-next-checkers (checker)
@@ -2497,6 +2504,21 @@ flycheck-buffer-saved-p
 
 
 ;;; Extending generic checkers
+(defun flycheck-remove-next-checker (checker next-checker)
+  "After CHECKER remove a NEXT-CHECKER.
+
+CHECKER is a syntax checker symbol, from which to remove NEXT
+checker.
+
+NEXT-CHECKER is a symbol denoting the syntax checker to not run after
+CHECKER any more."
+  (unless (flycheck-valid-checker-p checker)
+    (error "%s is not a valid syntax checker" checker))
+  (setf (flycheck-checker-get checker 'next-checkers)
+        (seq-filter (lambda (next) (not (eq (flycheck--get-next-checker-symbol next)
+                                       next-checker)))
+                    (flycheck-checker-get checker 'next-checkers))))
+
 (defun flycheck-add-next-checker (checker next &optional append)
   "After CHECKER add a NEXT checker.
 
@@ -2517,10 +2539,13 @@ flycheck-add-next-checker
   (unless (flycheck-valid-checker-p checker)
     (error "%s is not a valid syntax checker" checker))
   (flycheck-validate-next-checker next 'strict)
-  (if append
-      (setf (flycheck-checker-get checker 'next-checkers)
-            (append (flycheck-checker-get checker 'next-checkers) (list next)))
-    (push next (flycheck-checker-get checker 'next-checkers))))
+  (let ((next-checkers
+         (flycheck-remove-next-checker
+          (flycheck--get-next-checker-symbol next)
+          (flycheck-checker-get checker 'next-checkers))))
+    (setf (flycheck-checker-get checker 'next-checkers)
+          (if append (append next-checkers (list next))
+            (cons next next-checkers)))))
 
 (defun flycheck-add-mode (checker mode)
   "To CHECKER add a new major MODE.
```

File: flycheck.el [hunk 1]
```diff
@@ -2507,10 +2507,16 @@ APPEND is non-nil."
...
+         (exists-next-checkers (flycheck-checker-get checker 'next-checkers))
+         (exists-next-checkers-sym (mapcar get-symbol exists-next-checkers)))
+    (if (memq nextsym exists-next-checkers-sym)
+        exists-next-checkers
       (setf (flycheck-checker-get checker 'next-checkers)
-            (append (flycheck-checker-get checker 'next-checkers) (list next)))
-    (push next (flycheck-checker-get checker 'next-checkers))))
+            (if append
+                (append exists-next-checkers (list next))
+              (append (list next) exists-next-checkers))))))
```
Comment (cpitclaudel) [inline flycheck.el]: Doesn't this code lose levels?

**Comment by conao3** [comment]
It has been a long time since I submitted this patch and it is difficult to understand the context.
If your patch solves the problem, please commit the patch to `master`.

**Comment by cpitclaudel** [comment]
OK, I will do that, with proper credit.

**Comment by cpitclaudel** [comment]
I think this is ready.  I've added tests and preserved authorship on the commits.  @fmdkdd, would you like to have a look?

**Comment by cpitclaudel** [comment]
Thanks a lot!

**Comment by conao3** [comment]
Thanks also from me!

**Pull request merged**